From fb4ee96ece361017d6035743bf1808eb3622da60 Mon Sep 17 00:00:00 2001
From: Sridharan S N <quic_sridsn@quicinc.com>
Date: Mon, 15 Apr 2024 07:25:33 +0530
Subject: [PATCH] drivers: firmware: Add delay before doing psci reset and
 poweroff

Adding delay to ensure all the uart buffer contents
are pushed to console before reset, poweroff or halt.

Cherry-pick of the below 2 commits:
32f8d4880226 ("drivers: firmware: Add 100 ms delay before doing psci reset")
568870ef8c8b ("drivers: firmware: Add 10 ms delay before doing psci poweroff")

Change-Id: I60b7f976a716daf3b2a80381b3cf0ad83980e697
Signed-off-by: Sridharan S N <quic_sridsn@quicinc.com>
---
 drivers/firmware/psci/psci.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/drivers/firmware/psci/psci.c b/drivers/firmware/psci/psci.c
index f78249fe2512..dab429a6f242 100644
--- a/drivers/firmware/psci/psci.c
+++ b/drivers/firmware/psci/psci.c
@@ -19,6 +19,7 @@
 #include <linux/reboot.h>
 #include <linux/slab.h>
 #include <linux/suspend.h>
+#include <linux/delay.h>
 
 #include <uapi/linux/psci.h>
 
@@ -307,6 +308,31 @@ static int get_set_conduit_method(const struct device_node *np)
 static int psci_sys_reset(struct notifier_block *nb, unsigned long action,
 			  void *data)
 {
+	/*
+	 * From kernel, psci_sys_reset function is called during reboot for
+	 * both warm and cold reset.
+	 *
+	 * QSEE differentiates between warm and cold reset by checking the
+	 * TZ_WONCE register, which is set by SBL based on CDT settings.
+	 *
+	 * During cold reset, QSEE communicates with RPM via glink and requests
+	 * RPM to reset and during warm reset, QSEE directly toggles PSHOLD
+	 * register.
+	 *
+	 * During warm reset case, the "Restarting system" print after reboot
+	 * command is given is not flushed from the serial driver to console
+	 * before which the board gets rebooted.
+	 *
+	 * The pr_emerg("Restarting system\n") is called after the reboot
+	 * notifier callbacks are finished, therefore even if we add a
+	 * reboot notifier in serial driver and flush the tx buffer, it still
+	 * wouldn't help.
+	 *
+	 * We have added this 100 ms delay before invoking psci call for the
+	 * serial driver to flush the tx buf contents to console before
+	 * rebooting.
+	 */
+	mdelay(100);
 	if ((reboot_mode == REBOOT_WARM || reboot_mode == REBOOT_SOFT) &&
 	    psci_system_reset2_supported) {
 		/*
@@ -329,6 +355,7 @@ static struct notifier_block psci_sys_reset_nb = {
 
 static void psci_sys_poweroff(void)
 {
+	mdelay(10);
 	invoke_psci_fn(PSCI_0_2_FN_SYSTEM_OFF, 0, 0, 0);
 }
 
-- 
2.34.1

